app/main.py

```diff
@@ -1,2 +1,125 @@
+import copy
+import pprint
+from collections.abc import Hashable
+from typing import Union, Any, Iterator
+
+
 class Dictionary:
-    pass
+    def __init__(self) -> None:
+        self.hash_table = [[] for _ in range(8)]
+        self.collisions = []
+        self.old_hash_table = None
+        self.capacity = 8
+
+    def resize(self) -> None:
+        if not len(self) / self.capacity > 2 / 3:
+            return
+
+        """Resize"""
+        self.capacity *= 2
+        self.old_hash_table = copy.deepcopy(self.hash_table)
+        self.hash_table = [[] for bucket in range(self.capacity)]
+        self.collisions.clear()
+
+        for bucket in self.old_hash_table:
+            if len(bucket):
+                self.__setitem__(key=bucket[0], value=bucket[1])
+
+    def __setitem__(self, key: Hashable, value: Any) -> None:
+        if not isinstance(key, Hashable):
+            raise TypeError(f"unhashable type: '{type(key).__name__}'")
+
+        index = hash(key) % self.capacity
+        item = [key, value, hash(key)]
+
+        """Empty bucket"""
+        if not len(self.hash_table[index]):
+            self.hash_table[index] = item
+            self.resize()
+            return
+
+        """Reassign value"""
+        if self.hash_table[index][0] == key:
+            self.hash_table[index][1] = value
+            return
+
+        """Check old collisions"""
+        for collision in self.collisions:
+            if collision[0] == key:
+                collision_item_index = collision[1]
+                self.hash_table[collision_item_index][1] = value
+                return
```

Comment:
This is a little strange. Just find the first empty cell

Reply:
The same in getitem

Reply:
And delitem

Reply:
"The method def __setitem__ first tries to:

1. Find an empty bucket and fill it with the value (+return)
2. Find a bucket with the same key and overwrite its value (+return)
In this case, I handle the scenario where I try to overwrite the value of a bucket that had a collision during placement. To find the correct bucket, I access self.collisions to get its index and then change the value in self.hash_table using the index from the collisions list.

That's why I cannot search for an empty cell to write the value, as this function is **already** performed in the code blocks above, namely in the "Empty bucket" and "Reassign value" parts, which are **protected by return** statements as well.

Reply:
But it makes code harder. You can avoid the second list for this task. But If you want you can keep it